readme
su
make
make install
make uninstall


debian/changelog
intel-classmate-function-keys (1.0-2) testing; urgency=low

  * Autostart de /usr/bin/fnkey/fnkey

 -- Miguel Angel Garcia <miguel.garcia@gmail.com>  Wed, 04 Sep 2013 17:20:09 -0300

intel-classmate-function-keys (1.0) testing; urgency=low

  * Initial release.

 -- Miguel Angel Garcia <miguel.garcia@gmail.com>  Fri, 16 Aug 2013 17:09:11 -0300


button.h
/*
* Button header file
*
* Copyright (C) 2009  Intel Corporation
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


#ifndef __BUTTON_H
#define __BUTTON_H

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <time.h>
#include <errno.h>
#include <sys/time.h>
#include <signal.h>
#include <sys/resource.h>
#include <sys/wait.h> 
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xlibint.h>
#include <X11/extensions/XTest.h>
#include <string.h>
#include <unistd.h>

typedef enum{
        KEY_PRESSED = KeyPressMask,
        KEY_RELEASED = KeyReleaseMask,
} KEY_STATUS_MASK;

/* Button.h */
Window get_active_Xid(Display *xdpy);
int set_window_focus(Display *xdpy, Window Xid);
int set_window_active(Display *xdpy, Window Xid);
Window search_wid_from_name(Display *xdpy, char *name);
void send_fake_key(Display *xdpy, Window Xid, int keycode, KEY_STATUS_MASK mask, int modstate);
void pageup(Display *xdpy, KEY_STATUS_MASK mask);
void pagedown(Display *xdpy, KEY_STATUS_MASK mask);
void launch_camera(Display *xdpy);
void capture_photo(Display *xdpy, KEY_STATUS_MASK mask);
void handle_button_msg(Display *xdpy, int key_msg);

#endif

/*END OF FILE*/



cmpc_pm.h
/*
* Copyright (C) 2009  Intel Corporation
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#define ACPI_MAX_STRING 80

typedef struct {
	unsigned int		version;
	unsigned long long	status;
	char			pathname[ACPI_MAX_STRING];
	unsigned int		inputparam;
} cmpc_pm_acpi_t;

#define CMPC_PM_ACPI_MAGIC 'A'

/* Clear all state associated w/ device
 *  input - none
 *  output - none
 */
#define CMPC_PM_ACPI_CLEAR			_IOW(CMPC_PM_ACPI_MAGIC, 0, cmpc_pm_acpi_t)

/* Return success if pathname exists
 *  input - pathname
 *  output - none
 */
#define CMPC_PM_ACPI_EXISTS			_IOW(CMPC_PM_ACPI_MAGIC, 1, cmpc_pm_acpi_t)

/* Return type of object
 *  input - pathname
 *  output - status = type
 */
#define CMPC_PM_ACPI_GET_TYPE		_IOWR(CMPC_PM_ACPI_MAGIC, 2, cmpc_pm_acpi_t)

/* Evaluate an object
 *  input - pathname, write buffer = arg list
 *  output - status = length of read buffer
 *           read buffer = eval data
 */
#define CMPC_PM_ACPI_EVALUATE_OBJ		_IOWR(CMPC_PM_ACPI_MAGIC, 3, cmpc_pm_acpi_t)

/* Get Next objects
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = list of child objects
 */
#define CMPC_PM_ACPI_GET_NEXT		_IOWR(CMPC_PM_ACPI_MAGIC, 4, cmpc_pm_acpi_t)

/* Get Next objects
 *  input - pathname = PNP _HID/_CID value to look for
 *  output - status = length of read buffer
 *           read buffer = list of matching devices
 */
#define CMPC_PM_ACPI_GET_DEVICES		_IOWR(CMPC_PM_ACPI_MAGIC, 5, cmpc_pm_acpi_t)

/* Get object info structure
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = object info structure
 */
#define CMPC_PM_ACPI_GET_OBJ_INFO		_IOWR(CMPC_PM_ACPI_MAGIC, 6, cmpc_pm_acpi_t)

/* Get parent object
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = path of parent object
 */
#define CMPC_PM_ACPI_GET_PARENT		_IOWR(CMPC_PM_ACPI_MAGIC, 7, cmpc_pm_acpi_t)

/* Get brigtness of LCD
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = path of parent object
 */
#define CMPC_PM_ACPI_GET_BRIGHTNESS		_IOWR(CMPC_PM_ACPI_MAGIC, 8, cmpc_pm_acpi_t)

/* Get status of Wireless
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = path of parent object
 */
#define CMPC_PM_ACPI_GET_WIRELESS 	_IOWR(CMPC_PM_ACPI_MAGIC, 9, cmpc_pm_acpi_t)

/* Get status of LAN Device
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = path of parent object
 */
#define CMPC_PM_ACPI_GET_LAN 	_IOWR(CMPC_PM_ACPI_MAGIC, 10, cmpc_pm_acpi_t)

/* Get status of Card reader
 *  input - pathname
 *  output - status = length of read buffer
 *           read buffer = path of parent object
 */
#define CMPC_PM_ACPI_GET_CARDREADER 	_IOWR(CMPC_PM_ACPI_MAGIC, 11, cmpc_pm_acpi_t)

/* Set brigtness of LCD
 *  input - pathname
            inputparam = value to set
 */
#define CMPC_PM_ACPI_SET_BRIGHTNESS		_IOWR(CMPC_PM_ACPI_MAGIC, 12, cmpc_pm_acpi_t)

/* Enable/disable the wireless device
 *  input - pathname
            inputparam = value to set
 */
#define CMPC_PM_ACPI_SET_WIRELESS		_IOWR(CMPC_PM_ACPI_MAGIC, 13, cmpc_pm_acpi_t)

/* Enable/disable the LAN Device
 *  input - pathname
            inputparam = value to set
 */
#define CMPC_PM_ACPI_SET_LAN		_IOWR(CMPC_PM_ACPI_MAGIC, 14, cmpc_pm_acpi_t)

/* Enable/disable the card reader
 *  input - pathname
            inputparam = value to set
 */
#define CMPC_PM_ACPI_SET_CARDREADER		_IOWR(CMPC_PM_ACPI_MAGIC, 15, cmpc_pm_acpi_t)


OnScrDsp.h
/*
* Copyright (C) 2009  Intel Corporation
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License, version 2, as
* published by the Free Software Foundation.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#ifndef ON_SCR_DSP_H
#define ON_SCR_DSP_H

#include <assert.h>
#include <stdio.h>
#include <gtk/gtk.h>


/**
 * @class OnScrDsp
 *
 * Display status of WLAN, volume, display, brightness, Caps Lock, Scroll
 * Lock and Num Lock.
 */
class OnScrDsp
{
public:
	
	typedef enum {
		DSP_NONE,
		DSP_LCD,
		DSP_VGA,
		DSP_DUAL,
	} DSP_STA;

	typedef enum {
		WND_BRG,
		WND_DSP,
		WND_WRL,
	} WND_ID;

	/**
	 * Get the single instance of OnScrDsp.
	 *
	 * @return instance of OnScrDsp
	 */
	static OnScrDsp &GetIns()
	{
		static OnScrDsp osd;
		return osd;
	};

	void ShowWnd(WND_ID id, int lvl);

private:
	OnScrDsp();
	~OnScrDsp();

	/* declare copy constructor and equal operator to prevent
	 * copied instance
	 */
	OnScrDsp(const OnScrDsp &);
	OnScrDsp &operator =(const OnScrDsp &);

	void SetWnd(GtkWidget **pWnd, GCallback gcb, gint wdt, gint hgh, 
		gint x, gint y);
	void HideWnd();
	void DrawBrg();
	void DrawBrgLvl();
	void DrawCps();
	void DrawDsp();
	void DrawMut();
	void DrawNum();
	void DrawPrv();
	void DrawScr();
	void DrawVlm();
	void DrawVlmLvl();
	void DrawWrl();

	GdkPixbuf *m_pBrgBckGrn; /**< image of brightness background */
	GdkPixbuf *m_pBrgLvl; /**< image of brightness level */
	GdkPixbuf *m_pBrgLvlBln; /**< image of blank brightness level */
	GdkPixbuf *m_pDspDual; /**< image of dual display */
	GdkPixbuf *m_pDspVga; /**< image of external display */
	GdkPixbuf *m_pDspLcd; /**< image of internal LCD display */
	GdkPixbuf *m_pWrlOff; /**< image of wireless off */
	GdkPixbuf *m_pWrlOn; /**< image of wireless on */

	GtkWidget *m_pWnd[WND_WRL]; /**< window pointers from WND_BRG to WND_WRL */
	
	GdkScreen *m_pScr; /**< pointer to screen */

	GtkWidget *m_pCrrWnd; /**< current showing window */
	gint m_CrrTmr; /**< current timer id */
	gint m_CrrLvl; /**< current level of brightness or volume */

	static gint WDT_SCR; /**< width of screen */
	static gint HGH_SCR; /**< height of screen */

	static const gint HGH_BTT; /**< heihgt of bottom */
	static const gint WDT_RGH; /**< width of right margin */

	static const gint WDT_VLM; /**< width of brightness and volume images */
	static const gint HGH_VLM; /**< height of brightness and volume images */
	static const gint WDT_DSP; /**< width of display images */
	static const gint HGH_DSP; /**< height of display images */
	static const gint WDT_LCK; /**< width of Caps, Num, Scroll Lock images */ 
	static const gint HGH_LCK; /**< height of Caps, Num, Scroll Lock images */
	static const gint WDT_BRG_LVL; /**< width of brightness level image */
	static const gint HGH_BRG_LVL; /**< height of brightness level image */
	static const gint WDT_VLM_LVL; /**< width of brightness level image */
	static const gint HGH_VLM_LVL; /**< height of brightness level image */

	static const gint BRG_LVL_X; /**< x offset of brightness level */
	static const gint BRG_LVL_Y; /**< y offset of brightnes level */

	static const gint VLM_LVL_X; /**< x offset of volume level */
	static const gint VLM_LVL_Y; /**< y offset of volume level */

	static const gint TIMEOUT; /**< timeout to hide window */

	static const gint BRG_MAX_LVL; /**< max level of brightness */
	static const gint VLM_MAX_LVL; /**< max level of volume */

	static gint VLM_X; /**< x position of brightness and volume images */
	static gint VLM_Y; /**< y position of brightness and volume images */
	static gint CPS_X; /**< x position of Caps Lock images */
	static gint CPS_Y; /**< y position of Caps Lock images */
	static gint DSP_X; /**< x position of display images */
	static gint DSP_Y; /**< y position of display images */
	static gint NUM_X; /**< x position of Num Lock images */
	static gint NUM_Y; /**< y position of Num Lock images */
	static gint PRV_X; /**< x position of No Privilege images */
	static gint PRV_Y; /**< y position of No Privilege images */
	static gint SCR_X; /**< x position of Scroll Lock images */
	static gint SCR_Y; /**< y position of Scroll Lock images */
	static gint WRL_X; /**< x position of wireless images */
	static gint WRL_Y; /**< y position of wireless images */

	friend gboolean TmrHlp(gpointer);

	friend gboolean BrgExp(GtkWidget *, GdkEventExpose *, gpointer);
	friend gboolean DspExp(GtkWidget *, GdkEventExpose *, gpointer);
	friend gboolean WrlExp(GtkWidget *, GdkEventExpose *, gpointer);
};
#endif /* ON_SCR_DSP_H */


OnScrDsp.cpp
/*
* Copyright (C) 2009  Intel Corporation
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License, version 2, as
* published by the Free Software Foundation.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "OnScrDsp.h"

gint OnScrDsp::WDT_SCR = 1024;
gint OnScrDsp::HGH_SCR = 600;

const gint OnScrDsp::HGH_BTT = 45;
const gint OnScrDsp::WDT_RGH = 12;

const gint OnScrDsp::WDT_VLM = 250;
const gint OnScrDsp::HGH_VLM = 60;
const gint OnScrDsp::WDT_DSP = 102;
const gint OnScrDsp::HGH_DSP = 49;
const gint OnScrDsp::WDT_LCK = 49;
const gint OnScrDsp::HGH_LCK = 49;
const gint OnScrDsp::WDT_BRG_LVL = 27;
const gint OnScrDsp::HGH_BRG_LVL = 38;
const gint OnScrDsp::WDT_VLM_LVL = 14;
const gint OnScrDsp::HGH_VLM_LVL = 36;

const gint OnScrDsp::BRG_LVL_X = 46;
const gint OnScrDsp::BRG_LVL_Y = 12;
const gint OnScrDsp::VLM_LVL_X = 40;
const gint OnScrDsp::VLM_LVL_Y = 12;

const gint OnScrDsp::TIMEOUT = 5000;

const gint OnScrDsp::BRG_MAX_LVL = 7;
const gint OnScrDsp::VLM_MAX_LVL = 14;

gint OnScrDsp::VLM_X = (WDT_SCR - WDT_VLM) / 2;
gint OnScrDsp::VLM_Y = HGH_SCR - HGH_BTT - HGH_VLM;
gint OnScrDsp::CPS_X = WDT_SCR - WDT_LCK - WDT_RGH;
gint OnScrDsp::CPS_Y = HGH_SCR - HGH_BTT - HGH_LCK;
gint OnScrDsp::DSP_X = (WDT_SCR - WDT_DSP) / 2;
gint OnScrDsp::DSP_Y = HGH_SCR - HGH_BTT - HGH_DSP;
gint OnScrDsp::WRL_X = OnScrDsp::CPS_X;
gint OnScrDsp::WRL_Y = OnScrDsp::CPS_Y;

/**
 * helper to handler timeout
 */
gboolean TmrHlp(gpointer)
{
	OnScrDsp::GetIns().HideWnd();
	return FALSE;
}

/**
 * expose event helper for brightness.
 */
gboolean BrgExp(GtkWidget *, GdkEventExpose *, gpointer)
{
	OnScrDsp::GetIns().DrawBrg();
	return TRUE;
}

/**
 * expose event helper for display.
 */
gboolean DspExp(GtkWidget *, GdkEventExpose *, gpointer)
{
	OnScrDsp::GetIns().DrawDsp();
	return TRUE;
}

/**
 * expose event helper for wireless.
 */
gboolean WrlExp(GtkWidget *, GdkEventExpose *, gpointer)
{
	OnScrDsp::GetIns().DrawWrl();
	return TRUE;
}

/**
 * Constructor of OnScrDsp.
 */
OnScrDsp::OnScrDsp()
	: m_pBrgBckGrn(NULL), m_pBrgLvl(NULL), m_pBrgLvlBln(NULL),
	m_pDspDual(NULL), m_pDspVga(NULL), m_pDspLcd(NULL),
	m_pWrlOff(NULL), m_pWrlOn(NULL),
	m_pCrrWnd(NULL), m_CrrTmr(0), m_CrrLvl(0)
{
	int i = 0;

	m_pBrgBckGrn = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Brightness_BG.png", NULL);
	m_pBrgLvl = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Brightness_Level.png", NULL);
	m_pBrgLvlBln = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Brightness_Level_Blank.png", NULL);
	m_pDspDual = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Display_Dual.png", NULL);
	m_pDspVga = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Display_External.png", NULL);
	m_pDspLcd = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Display_LCD.png", NULL);
	m_pWrlOff = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Wireless_Off.png", NULL);
	m_pWrlOn = gdk_pixbuf_new_from_file(
		"/usr/bin/fnkey/Images/Wireless_On.png", NULL);

	assert(m_pBrgBckGrn && m_pBrgLvl && m_pBrgLvlBln &&
		m_pDspDual && m_pDspVga&& m_pDspLcd &&
		m_pWrlOff && m_pWrlOn);

	for (i = 0; i <= WND_WRL; i++) {
		m_pWnd[i] = gtk_window_new(GTK_WINDOW_POPUP);
		assert(m_pWnd[i]);
	}

	SetWnd(&m_pWnd[WND_BRG], GCallback(BrgExp), WDT_VLM, HGH_VLM, VLM_X, VLM_Y);
	SetWnd(&m_pWnd[WND_DSP], GCallback(DspExp), WDT_DSP, HGH_DSP, DSP_X, DSP_Y);
	SetWnd(&m_pWnd[WND_WRL], GCallback(WrlExp), WDT_LCK, HGH_LCK, WRL_X, WRL_Y);
}


/**
 * Destructor of OnScrDsp.
 */
OnScrDsp::~OnScrDsp()
{

}

/**
 * Set style of a window.
 *
 * @param pWnd	GtkWindow to set style
 */
void OnScrDsp::SetWnd(GtkWidget **pWnd, GCallback gcb, gint wdt, gint hgh,
	  gint x, gint y)
{
	gtk_widget_set_app_paintable(*pWnd, TRUE);
	gtk_widget_set_double_buffered(*pWnd, FALSE);
	gtk_widget_set_size_request(*pWnd, wdt, hgh);
	gtk_window_move((GtkWindow *)(*pWnd), x, y);
	gtk_widget_realize(*pWnd);
	gdk_window_set_back_pixmap((*pWnd)->window, NULL, FALSE);


	g_signal_connect(G_OBJECT(*pWnd), "destroy",
		G_CALLBACK(gtk_main_quit), NULL);
	g_signal_connect(G_OBJECT(*pWnd), "expose_event", gcb, NULL);

	gtk_widget_add_events(*pWnd, GDK_EXPOSURE_MASK);
}

/**
 * Show brightness window.
 *
 * @param id	window ID
 * @param lvl	brightness level
 */
void OnScrDsp::ShowWnd(WND_ID id, int lvl)
{
	GdkScreen *scr;

	g_usleep(10);
	gdk_threads_enter();

	if (m_CrrTmr) {
		g_source_remove(m_CrrTmr);
		m_CrrTmr = 0;
	}

	/* to avoid flicker, reduce hide and show */
	if (m_pCrrWnd == m_pWnd[id]) {
		switch (id) {
		case WND_BRG:
			m_CrrLvl = lvl;
			DrawBrgLvl();
			goto AFTER_SHOW;
		case WND_DSP:
		case WND_WRL:
		default:
			break;
			/* if not match, continue hide and show */
		}
	} 
	if (m_pCrrWnd)
		gtk_widget_hide(m_pCrrWnd);

	/* get current screen resolution to recalc icon's position */
	scr = gdk_screen_get_default();
	HGH_SCR = gdk_screen_get_height(scr);
	WDT_SCR = gdk_screen_get_width(scr);

	switch (id) {
	case WND_BRG:
		VLM_X = (WDT_SCR - WDT_VLM) / 2;
		VLM_Y = HGH_SCR - HGH_BTT - HGH_VLM;
		gtk_window_move((GtkWindow *)m_pWnd[id], VLM_X, VLM_Y);
		break;
	case WND_DSP:
		DSP_X = (WDT_SCR - WDT_DSP) / 2;
		DSP_Y = HGH_SCR - HGH_BTT - HGH_DSP;
		gtk_window_move((GtkWindow *)m_pWnd[id], DSP_X, DSP_Y);
		break;
	case WND_WRL:
		WRL_X = WDT_SCR - WDT_LCK - WDT_RGH;
		WRL_Y = HGH_SCR - HGH_BTT - HGH_LCK;
		gtk_window_move((GtkWindow *)m_pWnd[id], WRL_X, WRL_Y);
		break;
	}

        m_pCrrWnd = m_pWnd[id];
        m_CrrLvl = lvl;
	gtk_widget_show(m_pCrrWnd);
	gtk_widget_hide(m_pCrrWnd);
	gtk_widget_show(m_pCrrWnd);

AFTER_SHOW:
	m_CrrTmr = g_timeout_add(TIMEOUT, &TmrHlp, NULL);
	gdk_threads_leave();
	g_usleep(10);
}

/**
 * Hide current window
 */
void OnScrDsp::HideWnd()
{
	g_usleep(10);
	gdk_threads_enter();
	gtk_widget_hide(m_pCrrWnd);
	m_CrrLvl = 0;
	m_pCrrWnd = NULL;
	gdk_threads_leave();
	g_usleep(10);
}

/**
 * draw brightness window
 */
void OnScrDsp::DrawBrg()
{
	gdk_draw_pixbuf(m_pWnd[WND_BRG]->window, NULL, m_pBrgBckGrn,
		0, 0, 0, 0, WDT_VLM, HGH_VLM, GDK_RGB_DITHER_NORMAL, 0, 0);
	DrawBrgLvl();
}

/**
 * draw brighness level
 */
void OnScrDsp::DrawBrgLvl()
{
	int i;

	/* check if the brightness value is correct */
	if (m_CrrLvl > BRG_MAX_LVL) {
		m_CrrLvl = BRG_MAX_LVL;
	} else if (m_CrrLvl < 0) {
		m_CrrLvl = 0;
	}

	/* draw level */
	for (i = 0; i < m_CrrLvl; i++) {
		gdk_draw_pixbuf(m_pWnd[WND_BRG]->window, NULL, m_pBrgLvl,
			0, 0, BRG_LVL_X + WDT_BRG_LVL * i, BRG_LVL_Y, WDT_BRG_LVL,
			HGH_BRG_LVL, GDK_RGB_DITHER_NORMAL, 0, 0);
	}

	/* draw blank level */
	for (i = m_CrrLvl; i < BRG_MAX_LVL; i++) {
		gdk_draw_pixbuf(m_pWnd[WND_BRG]->window, NULL, m_pBrgLvlBln,
			0, 0, BRG_LVL_X + WDT_BRG_LVL * i, BRG_LVL_Y, WDT_BRG_LVL,
			HGH_BRG_LVL, GDK_RGB_DITHER_NORMAL, 0, 0);
	}
}

/**
 * draw display window
 */
void OnScrDsp::DrawDsp()
{
	if (DSP_LCD == m_CrrLvl) {
		gdk_draw_pixbuf(m_pWnd[WND_DSP]->window, NULL, m_pDspLcd,
			0, 0, 0, 0, WDT_DSP, HGH_DSP, GDK_RGB_DITHER_NORMAL, 0, 0);
	} else if (DSP_VGA == m_CrrLvl) {
		gdk_draw_pixbuf(m_pWnd[WND_DSP]->window, NULL, m_pDspVga,
			0, 0, 0, 0, WDT_DSP, HGH_DSP, GDK_RGB_DITHER_NORMAL, 0, 0);
	} else if (DSP_DUAL == m_CrrLvl) {
		gdk_draw_pixbuf(m_pWnd[WND_DSP]->window, NULL, m_pDspDual,
			0, 0, 0, 0, WDT_DSP, HGH_DSP, GDK_RGB_DITHER_NORMAL, 0, 0);
	} else if (DSP_NONE == m_CrrLvl) {
		return;
	}
}

/**
 * draw wireless window
 */
void OnScrDsp::DrawWrl()
{
	if (m_CrrLvl) {
		gdk_draw_pixbuf(m_pWnd[WND_WRL]->window, NULL, m_pWrlOn,
			0, 0, 0, 0, WDT_LCK, HGH_LCK, GDK_RGB_DITHER_NORMAL, 0, 0);
	} else {
		gdk_draw_pixbuf(m_pWnd[WND_WRL]->window, NULL, m_pWrlOff,
			0, 0, 0, 0, WDT_LCK, HGH_LCK, GDK_RGB_DITHER_NORMAL, 0, 0);
	}
}
